service/src/main/java/uk/gov/hmcts/reform/fpl/service/ConfidentialDetailsService.java

```diff
@@ -3,31 +3,118 @@
+    }
+
+    public <T extends ConfidentialParty<T>> List<Element<T>> removeConfidentialDetails(List<Element<T>> details) {
+        return details.stream()
+            .filter(element -> element.getValue() != null)
+            .map(item -> {
+                if (item.getValue().containsConfidentialDetails()) {
+                    return element(item.getId(), item.getValue().removeConfidentialDetails());
+                } else {
+                    return item;
+                }
+            })
             .collect(toList());
     }
 
-    public <T> void addConfidentialDetailsToCaseDetails(CaseDetails caseDetails,
-                                                        List<Element<T>> confidentialDetails,
-                                                        ConfidentialPartyType type) {
+    public <T> void addConfidentialDetailsToCase(CaseDetails caseDetails,
+                                                 List<Element<T>> confidentialDetails,
+                                                 ConfidentialPartyType type) {
         if (isNotEmpty(confidentialDetails)) {
             caseDetails.getData().put(type.getCaseDataKey(), confidentialDetails);
         } else {
             caseDetails.getData().remove(type.getCaseDataKey());
         }
     }
+
+    private Child expandedChild() {
+        return Child.builder().build().expandCollection();
+    }
+
+    private Respondent expandedRespondent() {
+        return Respondent.builder().build().expandCollection();
+    }
+
+    private <T extends ConfidentialParty<T>> List<Element<T>> prepareCollection(List<Element<T>> all,
+                                                                                List<Element<T>> confidential,
+                                                                                T defaultValue) {
+        List<Element<T>> collection = new ArrayList<>();
+
+        if (all.isEmpty()) {
+            collection.add(element(defaultValue));
+
+        } else if (confidential.isEmpty()) {
+            return all;
+
+        } else {
+            all.forEach(element -> {
+                T party = element.getValue();
+                if (party.containsConfidentialDetails()) {
+                    T partyToAdd = getItemToAdd(confidential, element);
+                    T confidentialParty = party.addConfidentialDetails((partyToAdd.getConfidentialParty()));
+
+                    // code due to others following a different data structure.
+                    if (defaultValue == null) {
+                        List<UUID> ids = all.stream().map(Element::getId).collect(toList());
+
+                        confidentialParty = confidential.stream()
+                            .filter(other -> !ids.contains(other.getId()))
+                            .map(other -> party.addConfidentialDetails(other.getValue().getConfidentialParty()))
+                            .findFirst()
+                            .orElse(party);
+                    }
```

Comment:
could we extract this to separate method? This method doesn't read that badly, but this adds some mess :P 

Why we are doing this ids = all.stream()? 
We are already processing all elements in all collection, why equals to ID is not enough as in bottom method?

Reply:
because we haven't migrated others and we need to know which one is the first other. The first is the only one with an element id that will not match that in confidential. Horrid, I know 😢